File: v2/src/main/mongo/aggregate.ts
import { EJSON } from 'bson';
import type { MongoClient } from 'mongodb';
import type {
  AggregateRequest,
  AggregateResult,
} from '../../shared/aggregation.js';
import { parseFilter } from './queryParser.js';

const PAGE_FETCH_EXTRA = 1;

export async function executeAggregate(
  client: MongoClient,
  req: AggregateRequest,
): Promise<AggregateResult> {
  const t0 = performance.now();
  const limit = req.limit > 0 ? Math.floor(req.limit) : 50;

  const pipeline = req.pipeline.map(({ operator, body }) => {
    if (!operator.startsWith('$')) {
      throw new Error(`Stage operator must start with $: ${operator}`);
    }
    const parsedBody = parseFilter(body);
    return { [operator]: parsedBody } as Record<string, unknown>;
  });

  // Append a final limit so we don't pull more than we can show.
  pipeline.push({ $limit: limit + PAGE_FETCH_EXTRA });

  let cursor = client
    .db(req.db)
    .collection(req.collection)
    .aggregate(pipeline, { allowDiskUse: true });
  if (req.maxTimeMs && req.maxTimeMs > 0) {
    cursor = cursor.maxTimeMS(req.maxTimeMs);
  }

  const fetched = await cursor.toArray();
  const hasMore = fetched.length > limit;
  const rows = (hasMore ? fetched.slice(0, limit) : fetched).map((d) =>
    EJSON.stringify(d, { relaxed: false }),
  );

  return {
    ok: true,
    rows,
    hasMore,
    durationMs: Math.round(performance.now() - t0),
  };
}

File: v3/src/main/kotlin/io/mex/mongo/Aggregate.kt
package io.mex.mongo

import com.mongodb.client.MongoClient
import org.bson.Document
import org.bson.json.JsonMode
import org.bson.json.JsonWriterSettings
import java.util.concurrent.TimeUnit

data class AggregateRequest(
    val connectionId: String,
    val db: String,
    val collection: String,
    val pipeline: List<Pair<String, String>>, // operator → body
    val limit: Int = 50,
    val maxTimeMs: Long = 60_000,
    /** Documents to skip, appended as a `$skip` stage so results can be paged. */
    val skip: Int = 0,
)

private val EJSON_SETTINGS: JsonWriterSettings = JsonWriterSettings.builder()
    .outputMode(JsonMode.EXTENDED)
    .build()

fun executeAggregate(client: MongoClient, req: AggregateRequest): FindResult {
    val t0 = System.nanoTime()
    return runCatching {
        val limit = req.limit.coerceAtLeast(1)
        val user = req.pipeline.map { (op, body) ->
            require(op.startsWith("$")) { "Stage operator must start with $: $op" }
            Document(op, parseDocument(body))
        }
        // Paging stages go after the user's pipeline so they window its output.
        val paging = buildList {
            if (req.skip > 0) add(Document("\$skip", req.skip))
            add(Document("\$limit", limit + 1))
        }
        val stages = user + paging

        var cursor = client.getDatabase(req.db).getCollection(req.collection)
            .aggregate(stages)
        if (req.maxTimeMs > 0) cursor = cursor.maxTime(req.maxTimeMs, TimeUnit.MILLISECONDS)

        val docs = cursor.into(mutableListOf<Document>())
        val hasMore = docs.size > limit
        val rows = (if (hasMore) docs.dropLast(1) else docs).map { it.toJson(EJSON_SETTINGS) }
        FindResult.Ok(rows = rows, hasMore = hasMore, durationMs = (System.nanoTime() - t0) / 1_000_000)
    }.getOrElse {
        FindResult.Failed(it.message ?: it::class.java.simpleName, (System.nanoTime() - t0) / 1_000_000)
    }
}
